### DISCUSSION TITLE
Nested, async formatters, translation engine for generator

### DISCUSSION THREAD

**Discussion by 0x241F31** [ivanhofer/typesafe-i18n#293] [Ideas]
Hello. Thank you for great tool.
I'm thinking about using a random value in the formatter, but I'm more likely to get the values from the API. Is it possible to make the formatter asynchronous then?

Also , it could be useful to allow nested formatters, like: 
```ts
	const formatters: Formatters = {
		random: {
			// use it with {name|random.firstname}
			firstname: async () => await requestRandomApi(locale)
		}
	};
```

After switching from vscode i18-ally I would like to be able to automatically translate missing keys, for example, through a generator parameter. I tried to make it myself, but it is too complex because of need to build typescript AST. Probably I can help with it.

**Comment by ivanhofer** [comment]
@MrFoxPro thanks for you kind words.

## async formatter

Making formatters async would result in making the whole translation process async. So instead of
```svelte
<h1>{LL.my.key()}</h1>
```
you would need to write
```svelte
<h1>{await LL.my.key()}</h1>
```
everywhere. That's not ideal so I'm against that change.

For your example above you should probably not use a formatter, you should call this function in your template like this:
```tsx
import { locale } from './src/i18n/i18n-react' // import depends on the adapter used

const username = await requestRandomApi(locale)

<h1>{LL.my.key(username)}</h1>
```

if you still want to use a formatter, you could use an approach like this:
```ts
import type { FormattersInitializer } from 'typesafe-i18n'
import type { Formatters, Locales } from './i18n-types'

const randomFirstnames: string[] = []
let i = 0

// you would need to make sure you call this from outside before
// initializing `typesafe-i18n` to already have some random values in the array
export const loadNext100RandomFirstNames = async () =>
	randomFirstnames = [...randomFirstnames, ...await requestRandomApi(locale)]

const getRandomFirstname = (locale: Locales) => {
	i++
	if (i < randomFirstnames.length - 10) {
		// when you see you don't have that much values left,
		// trigger the loading of new in the background
		loadNext100RandomFirstNames()
	}

	return randomFirstnames[i]
}

export const initFormatters: FormattersInitializer<Locales, Formatters> = (locale: Locales) => {
	const formatters: Formatters = {
		randomFirstname: getRandomFirstname(locale)
	}

	return formatters
}
```
## nesting

I think nesting makes sense only when you have a lot of formatters. I haven't seen more than 10 formatters in a project yet, so I would assume it is not relly needed since it makes the implementation more complicated and adds some bytes to the bundle size.

## automatic translation

Take a look at the [`exporter`](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/exporter) and [`importer`](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/importer) functionality. You could use the [`runAfterGenerator`-hook](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/generator#runaftergenerator) that calls a script that reads alle the entries from your base translation file via the `exporter`, sends it to a service that translates the strings and then writes it to disk via the `importer`.

I would love to provide such a functionality out-of-the box. Maybe you want to help me out and write that script to include it in a PR?

**Reply by 0x241F31** [reply]
I'm working on it
